src/file/ShareDialog.tsx

```diff
@@ -1,17 +1,92 @@
-import { IModalProps, IStyleFunctionOrObject, Theme } from '@fluentui/react';
-import React from 'react';
+import {
+    DialogFooter,
+    IModalProps,
+    IStyle,
+    IStyleFunctionOrObject,
+    Label,
+    Pivot,
+    PivotItem,
+    PrimaryButton,
+    Spinner,
+    TextField,
+    Theme,
+    mergeStyleSets,
+} from '@fluentui/react';
+import React, { useEffect, useState } from 'react';
 import { BaseDialog, IBaseDialogStyles } from '../BaseDialog';
+import { useScene } from '../SceneProvider';
+import { sceneToText } from '../file';
+
+const classNames = mergeStyleSets({
+    tab: {
+        minHeight: 200,
+        display: 'grid',
+        gridTemplateRows: '1fr auto',
+        gridTemplateAreas: `
+            "content"
+            "footer"
+        `,
+    } as IStyle,
+    footer: {
+        gridArea: 'footer',
+    } as IStyle,
+});
+
+const copyIconProps = { iconName: 'Copy' };
 
 export const ShareDialog: React.FC<IModalProps> = (props) => {
     return (
         <BaseDialog headerText="Share" {...props} dialogStyles={dialogStyles}>
-            <p>TODO</p>
+            <Pivot>
+                <PivotItem headerText="Export String" className={classNames.tab}>
+                    <ShareText />
+                </PivotItem>
+            </Pivot>
         </BaseDialog>
     );
 };
 
 const dialogStyles: IStyleFunctionOrObject<Theme, IBaseDialogStyles> = {
     body: {
+        minWidth: 500,
         maxWidth: '70ch',
     },
 };
+
+const labelStyles = mergeStyleSets({
+    message: { transition: 'none', opacity: 1 } as IStyle,
+    hidden: { transition: 'opacity 0.5s ease-in-out', opacity: 0 } as IStyle,
+});
+
+const ShareText: React.FC = () => {
+    const { scene } = useScene();
+    const [data, setData] = useState<string | undefined>(undefined);
+    const [copyMessageVisible, setMessageVisibility] = useState(false);
+    const [timeout, setMessageTimeout] = useState<number>();
+
+    useEffect(() => {
+        (async () => setData(await sceneToText(scene)))();
+    }, [scene]);
+
+    const doCopyToClipboard = () => {
+        navigator.clipboard.writeText(data || '');
+        setMessageVisibility(true);
+        clearTimeout(timeout);
```

Comment:
I think we need to have an extra `useEffect` here to make sure that the timeout callback can't get called after this component is destroyed. IIRC, that looks something like

```ts
useEffect(() => {
    return () => clearTimeout(timeout);
}, [timeout]);
```

Reply:
It looks like that doesn't actually work. https://felixgerschau.com/react-hooks-settimeout/ suggests using a ref to store the timer instead so that it does work.

Reply:
Thank you for that blog - It does allow me to correct a few mistakes I've made here, specifically using `useState` for the timeout, instead of `useEffect`.